### DISCUSSION TITLE
Sentinel-2 Mask via Scene Classification Layer

### DISCUSSION THREAD

**Discussion by DanSchoppe** [developmentseed/titiler#533] [Q&A]
TLDR: Can I mask Sentinel-2 data (raster tiles & statistics) based on Sentinel-2 SCL data?

---

Hello! I'm accessing Sentinel-2 data via TiTiler and STAC records, both for raster tiles and statistics. I'm hoping to leverage Sentinel-2's scene classification layer data to omit pixels matching certain classifications (eg. clouds), both to make pixels transparent in raster tiles and to exclude pixels in statistics calculations.

Is there a way to include this "NODATA" / null sentiment in an expression? For example, a Sentinel-2 NDVI expression that excludes `CLOUD_HIGH_PROBABILITY` and `CLOUD_MEDIUM_PROBABILITY` might be something along the lines of (warning: does not work, `NaN` is not valid in this context):
```
where((SCL==9)|(SCL==8),NaN,(B08-B04)/(B08+B04));
```

I've tinkered with `NumExpr` to no avail. developmentseed/titiler/discussions/487 looks like a promising path forward (`where` clause returns NODATA value like -10000 in one of the conditions) but it sounds like this strategy might not actually be a great option:
> The expression applied after data reading and we have no way to know if the mask was created from nodata / mask / alpha band, which is why we can't update the Mask

@vincentsarago you teased some upcoming changes regarding `post_process` and `user_input` — are those useful for this type of situation?

Thanks as always for the great tooling and all that you do!
Dan

**Comment by vincentsarago** [comment]
Custom Algorithm work in on going over https://github.com/developmentseed/titiler/pull/542 if you are interested :-) 
I'll add more example and documentation tomorrow

**Reply by DanSchoppe** [reply]
Lovely! Am I understanding correctly that you're envisioning this to be stateful, where a client would "register" an algorithm before being able to reference it by name? If so, I'm guessing you have a technical limitation in mind that prevents callers from specifying an algorithm at request time?

**Reply by DanSchoppe** [reply]
Ahhh, I think I did misunderstand. Let me try again...

Looks like algorithms can be added *in source code* to a forked version of TiTiler, which will then be available (by name) for use in future requests.

**Reply by vincentsarago** [reply]
> Looks like algorithms can be added in source code to a forked version of TiTiler, which will then be available (by name) for use in future requests.

Yes, we don't want to introduce heavy dependency. We could do like rio-tiler's colormap and morecantile TMS where you point to a directory and we look for algorithm

**Reply by DanSchoppe** [reply]
Makes sense. Thanks for the response @vincentsarago.

In this issue's specific case, did you have in mind that a Sentinel-2 mask-from-SCL algorithm could be built?

Alternatively, I could see this problem being generalized as a desire to, as a caller, express how a mask should be computed. I'm thinking along the lines of expression with named assets. Call me crazy, but I might even imagine allowing the mask to operate on a different STAC or COG URL 🤷.

In this latter case, I'm not sure if there'd be a way to achieve a general masking feature into an "algorithm", or whether it's more of a fundamental feature request to TiTiler and rio-tiler.

**Reply by DanSchoppe** [reply]
@vincentsarago does the above comment make sense?

AFAIK there isn't any ability to specify mask parameters separately from raster parameters, but maybe there's value in such a thing. Is it correct that a Sentinel-2-specific "algorithm" could be a work around, but perhaps this discussion is hinting at a request for more sophisticated mask parameters in underlying libraries (rasterio and/or rio-tiler)?
